### ISSUE TITLE
Comfortable development workflow?

### ISSUE THREAD

**Issue by kindaro** [coq/coq#13328]
### I am not doing work because the build system is lame.

I have been attempting to contribute to Coq since a while _(#11910, #11911)_, but every time I come back to it, there is some frustrating problem with the build system that makes me sad enough to abandon the work. For example, [the last time I discovered that **unit tests are not really performed**][1] and have not been for who knows how long a time. _(Resolved in #12289.)_ What I mean to express by this preamble is that, while such issues may be avoided so habitually by the frequent contributors as not to even be apparent to them, the experience of building and hacking is of any quality but comfortable. I understand that there is never enough hands to perfect everything, and the build system is of the least priority. But perhaps **there is not enough hands because it is so unnecessarily difficult to get going** in the first place!

[1]: https://gitter.im/coq/coq?at=5eb713ff9f0c955d7db58015

### The build system is lame.

One thing that makes me sad is that there are two build systems that depend on each other in a subtle way. These systems are `make` and `dune`.
* I can run an individual test only with the `make` system.
* I can run `utop` in a sub-directory only with the `dune` system.

But in order to use `dune`, I must remove `make`'s build products!

    % dune utop
    Error: Multiple rules generated for _build/default/META.coq:
    - file present in source tree
    - <internal location>
    Hint: rm -f META.coq

This means that I must either run all tests or rebuild a large part of the project every time I make a smallest modification. Either takes a long time — as long as hours. This gives a poor experience and makes me sad.

#### For a wider view:

* Why should there even be two build systems at once, for as long as at least the half a year I am around?
* Why should the two officially supported build systems squabble over some build products that are so insignificant as to even be ignored by the version control?

This does not make any sense.

### Possible solutions.

* In a more modern language ecosystem, there would be a flexible and reliable test runner. Example: Haskell's `tasty`. We cannot have this because porting the current test suite anywhere would require manual rewrites that would introduce errors. Perhaps some parts of the test suite, such as the innumerable `success` and `failure` bunches, could be automatically put under a modern test runner? Is there even a modern test runner in the Ocaml ecosystem?
* For some reason, `make` must keep build products all over the project, and `dune` must complain about that and crash. Can either of this be solved?
* Or maybe I am missing some blessed, simple and beautiful work flow that everyone is using? Let us put it to a more prominent place in the developers' documentation!

### I really want to help, only give me leadership.

If a system is to work, there must be a vision and an architecture. Any system left to float by itself inevitably becomes a mess, which is what we have here. It is not efficient to fix small things here and there. My proposition: `make` must be deprecated as soon as possible, and `dune` should be taught to run individual tests with colourful output. If there is a way I can help make it happen, I have hours upon hours of free work to offer.

**Comment by ejgallego** [comment]
Actually you can run individual tests with dune, tho the setup is not perfect yet. Just call the makefile in the test suite after the dune build. Better support is planned, I can try to push my branch and ask for help.

Note that as of today, Dune is still experimental. Migration of large build systems doesn't happen overnight, and for better or worse, Coq's governance model doesn't have "a vision and an architecture", but it more of a federated nature.

I agree this is not efficient.

**Comment by kindaro** [comment]
@ejgallego  I know you have things in works, but is it documented somewhere so that I could use it without bothering you? I gather that `dune` and `make` are all mixed together, It is not possible to understand how the conglomerate is supposed to work without detailed understanding of both, so I cannot offer any meaningful contribution without first receiving some guidance. If there is something specific I can help with, you can always alert me by electronic mail, or even simply create an issue and assign me to it.

How long can `dune` remain _«experimental»_? It already builds stuff. Suppose you remove `make` altogether. Surely the `dune` build system would then be polished in a matter of days.

**Comment by ejgallego** [comment]
Main developer documentation is under `dev/doc/build-system.dune.md` IIRC.

**Comment by kindaro** [comment]
I have been reading that. I do not think it says anything about running individual tests.

**Comment by kindaro** [comment]
A trivial solution is to run `make`, stash the build products somewhere, then put them back when running tests and remove again when running `dune utop`. This is straightforward to program, and it does work — I made a pair of shell scripts that execute smoothly. But there are thousands of files in the stash — 6538 to be precise — so moving them around incurs a delay of about a minute. Even if it was instant, this solution is laughable.

I present my scripts as a proof of concept.

`stash_build_products.sh`

    stash_location='.make_build_products'
    mkdir -p "$stash_location"
    git clean --dry-run -X | cut -d ' ' -f 3 | while read x
    do
        if [ "${x%%/*}" == '_build' ] || [ "${x%%/*}" == "$stash_location" ] || [ "`basename "$x"`" == '.merlin' ]
        then  # Do not stash these files.
            true
        else
            x_dirname="`dirname "$x"`"
            mkdir -p "$stash_location/$x_dirname"
            mv "$x" "$stash_location/$x"
            echo -n '.'
        fi
    done
    echo

`restore_build_products.sh`

    stash_location='.make_build_products'
    find "$stash_location" -type f | while read x
    do
        x_dirname="`dirname "$x"`"
        mkdir -p "${x_dirname#"$stash_location/"}"
        cp -a "$x" "${x#"$stash_location/"}"
        echo -n '.'
    done
    echo

**Comment by Zimmi48** [comment]
Ideally, the test-suite would be ported to Dune. Emilio indeed has a branch which does this partially (IIUC) but in the meantime, the solution that has been found (by @SkySkimmer) has been to adapt the test-suite Makefile to be compatible with a Dune-built Coq (this may be undocumented).

The way to use this is to build Coq with Dune, then to `cd test-suite && make` (with appropriate `-j` option). Any call to the test-suite directly through Dune is a waste of time because it is not yet incremental.

**Comment by kindaro** [comment]
I opened ocaml/dune#3933 to find out if there may be a way to pacify `dune` with the presence of `make`'s build products in the source tree.

**Comment by Zimmi48** [comment]
I think that all core developers agree that we need to get rid of the legacy build system and improve whatever needs to be in the new one. The main blocker so far has been native compute support. This is coming soon (see ocaml/dune#3210). After this is in, I propose that we remove immediately all redundant legacy `Makefile` and we work on fixing Dune during the upcoming release cycle (only under the pressure of having to fix the remaining build system will we make enough progress).

**Comment by Zimmi48** [comment]
>I have hours upon hours of free work to offer.

As you have already guessed, that's very valuable to us. Are you planning to attend the upcoming [CUDW](https://coq.discourse.group/t/next-cudw-2020-11-30-to-2020-12-04-save-the-date/1074). This should be a good time to interact with other Coq developers, push the Dune migration, and find where you can be most useful.

**Comment by kindaro** [comment]
> … in the meantime, the solution that has been found (by @SkySkimmer) has been to adapt the test-suite Makefile to be compatible with a Dune-built Coq (this may be undocumented).
> 
> The way to use this is to build Coq with Dune, then to `cd test-suite && make` (with appropriate `-j` option). …

The way you suggest would not work with a specific test or a directory. See:

    % (cd test-suite && make -j7)
    TEST      success/CaseInClause.v 
    TEST      success/CasesDep.v 
    TEST      success/Cases.v 
    TEST      success/coercions.v 
    TEST      success/cofixtac.v 
    TEST      success/coindprim.v 
    TEST      success/CombinedScheme.v 
    CHECK     success/cofixtac.v
    CHECK     success/CombinedScheme.v
    TEST      success/CompatCurrentFlag.v  (-compat 8.13)
    ...

    % (cd test-suite && make -j7 unit-tests/clib/inteq.ml.log)
    OCAMLC  unit-tests/src/utest.mli
    ocamlc: unknown option '-package'.
    Usage: ocamlc <options> <files>
    ...

    % (cd test-suite && make -j7 unit-tests)
    OCAMLC  unit-tests/src/utest.mli
    ocamlc: unknown option '-package'.
    Usage: ocamlc <options> <files>
    ...

It is not realistic for me to run the whole suite, and besides, that is already done by `dune build` and `dune runtest`.

**Comment by Zimmi48** [comment]
>It is not compatible with test driven work flow to run the whole test suite every time

This way of doing is precisely supposed to avoid rebuilding the whole test suite every time. I am not sure what is this error you're facing when building unit tests but it has nothing to do with this.

**Comment by kindaro** [comment]
A more thorough check reveals that some tests can be run separately, while others cannot. Example that works:

    % (cd test-suite && make -j7 success/CaseInClause.v.log )
    TEST      success/CaseInClause.v 
    CHECK     success/CaseInClause.v

So, yes, there must be issues with the unit test part.

**Comment by Zimmi48** [comment]
FWIW, I get the same error when trying to build unit tests locally (with a Dune-built Coq). I haven't tried another configuration.

**Comment by ejgallego** [comment]
Personally I wouldn't worry about making a hybrid make (for OCaml objects) and dune setup working. The answer is simple: don't ever use make, as Theo mentions it will removed soon.

There is a good reason (good in the sense that it is justified, however it is more of an historical articact) that Dune forbids objects file in the tree.

**Comment by kindaro** [comment]
@ejgallego  Suppose we put all unit tests into a `OUnit` test tree that would be built into a monolithic executable. This solution would prevent trivial problems due to the compiler checking things and it is something I can do now, rather than waiting indefinitely for a `dune` based test runner to appear. Would you back it up and help me get it merged?

**Comment by ejgallego** [comment]
> @ejgallego  Suppose we put all unit tests into a `OUnit` test tree that would be built into a monolithic executable. This solution would have less bit rot and it is something I can do now, rather than waiting indefinitely for a `dune` based test runner to appear. Would you back it up and help me get it merged?

I'm not familiar enough with the details, that could be interesting indeed. One thing to consider tho is that once dune supports the test-suite better, it will know about individual tests, so tests are only re-run when the dependencies change.

IMHO it'd be a bit inconvenient if the ounit test tree does force a full rerun.

**Comment by kindaro** [comment]
I am not aware of any prior art in the direction of using a build system to automagically select what checks to run. Has it been done before somewhere?

If you are interested in a discussion, I have a few thoughts:

1. We already had an issue with tests not being discovered.
    * It seems to me that a build system is not suitable for such tasks, since there would be no obvious error when a report is missing some lines.
    * A monolithic executable takes a small bit of work up front to set up but guarantees that all checks will run by virtue of the compiler checking the script.

    So there is the benefit of reliability to a monolithic executable.

2. For example, my project involves changing `coqpp` source code. Does it affect the output of `*.v` files? Theoretically, yes. Does it mean that I should spend upwards of 10 minutes re-running the whole test suite for every small change? It seems more practical to select the desired checks to run by hand. Therefore:
    * Being able to select checks by hand is sufficient.
    * Being able to let the runner automagically decide what to run is not enough.

    So, the smallest complete feature set consists of being able to select checks by hand.

I think the architecture should be thought through. Currently it appears nebulous. Once there is a clear design, it would be feasible to split work and get things done.

**Comment by ppedrot** [comment]
> I am missing some blessed, simple and beautiful work flow that everyone is using?

Run the tests on the CI when submitting a PR. Works for me ©.

**Comment by ejgallego** [comment]
> I am not aware of any prior art in the direction of using a build system to automagically select what checks to run. Has it been done before somewhere?

I am very far from an expert on the matter, but indeed I have heard about a lot of research in this area, that's even a mature field. Think of something large like Google Chrome, for sure devs don't want to run millions of tests for each rebuild.

I'd prefer an actual expert to comment tho. But in Dune itself, tests work this way. There is nothing special, tests correctly declare their dependencies and are re-run when the deps change. This is essential as the test suite is pretty large.

Regarding your points 1 and 2, a solution like the one we have in mind, covers both cases. Actually, manually selection of tests is a very bad idea, the system should select what tests to run automatically.

**Comment by kindaro** [comment]
So you already see clearly what needs to be done. [As someone else says elsewhere:][1]

[1]: https://github.com/ocaml/dune/issues/3933#issuecomment-724959273

> Why don't you try porting the couple of unit tests you're interested to dune and see how that goes?

Is there a design document I may consult so that I can get an understanding of what needs to be done and begin porting tests to `dune` in a way that may later be merged? Where do I start?

**Comment by ejgallego** [comment]
That's called "cutoff" I believe; but indeed if you modify something close to the root, it makes sense to rerun everything. But I agree that having a dep cutoff is useful in a bunch of cases and hopefully we can support that. But I see that as a refinement.

> Is there a design document I may consult so that I can get an understanding of what needs to be done and begin porting tests to `dune` in a way that may later be merged? Where do I start?

I'll submit my work in progress ASAP, but that covers only the Coq part, not the OCaml unit tests. For OCaml unit tests there are a few approaches, dune has good support for expect and recently awesome one for cram. We should google a bit to see what's the preferred solution.

I think adding a `dune` test setup to `test-suite/unit-tests` could be a good start. Also, exploring some other packages such as alcotest could be interesting. Best supported in indeed inline test as many library writers in OCaml like the testing to be next to the code, but that's a personal preference.

**Comment by kindaro** [comment]
So, the outline is like this:

* Exclusively `dune runtest` will be used to run checks.
* Every single independently runnable check will be a separate build unit.
* The checks currently present will fall under:
    * The category of _«custom tests»_ if they are OCaml files.
    * The category of _«cram tests»_ if they are based on Coq files. _(I am not sure how this should work — would you generate a _«cram script»_ for every Coq file?)_

    — As per [the `dune` manual].

Do I understand you right?

The down sides that I anticipate and fear are that:
* It is not trivial to provide _«quality of life»_ features such as colourful or machine-readable reporting.
* Some relevant checks may not run due to dependencies not being wired correctly. How are you planning to wire the dependencies from source files to the relevant checks?
* A test run for a feature close to the root of the dependency tree would require to re-run almost all checks, unless the depth limit feature is implemented in `dune`. This means that in practice checks would have to be selected to run manually on a per-directory basis when real time response is desired. [The `dune` manual] makes an impression that only custom checks can be run by name, since other types of checks do not have an associated executable — this would mean that it would be impossible to run individual Coq based checks.

Can you clarify how these deficiencies will be mitigated or why they would not appear in the first place?

[the `dune` manual]: https://dune.readthedocs.io/en/stable/tests.html

**Comment by ejgallego** [comment]
> * Exclusively `dune runtest` will be used to run checks.

That's the top level command, and very convenient to be the main driver as it both provides the hygienic env and the test suite deps for free, but if you think we should plug another test-runner under it IMHO I'd be fine. I'm not expert with tests at all.

> There will be no other test runners in the picture, so as to ensure maximum granularity of dependencies from sources to checks

I think that for the .v part, at least that would be cool, but I don't think that's an absolute rule. It is an efficient one tho as we just delegate everything to dune and `dune runtest` twice will do nothing.

> * For example, if you want to have an option for generating machine-readable report, you would have to teach `dune` to provide that.

I think it already has some support for quite a few cases, for example output tests. Further improvement are planned, so Coq would get the improvement for free. In my experience testing Coq you really care to have quick re-tests , easy promotion, and it is usually a very few tests failing so that's OK.

> * The checks currently present will fall under

Actually I think most tests would be generated regular rules, using `diff` action for some, the rest just expected exit status.

> * It is not trivial to provide _«quality of life»_ features such as colourful or machine-readable reporting.

I think the current quality of life status of that is not bad at all.

> * Some relevant checks may not run due to dependencies not being wired correctly. How are you planning to wire the dependencies from source files to the relevant checks?

coqdep 

> * This means that in practice checks would have to be selected to run manually on a per-directory basis when real time response is desired.

Test can be named for manual selection.

**Comment by ejgallego** [comment]
It is true tho that `coqc` is a really bad node in this setup, dependency-wise.

**Comment by kindaro** [comment]
#### I think I mostly understand you. There are some parts that I have questions about.

> > * The checks currently present will fall under
> 
> Actually I think most tests would be generated regular rules, using `diff` action for some, the rest just expected exit status.

Can you explain at some more length, maybe in such terms that I could look up in `dune`'s manual?

> > * Some relevant checks may not run due to dependencies not being wired correctly. How are you planning to wire the dependencies from source files to the relevant checks?
> 
> coqdep

This would not work for OCaml unit tests. How would that be handled? For example, what would `test-suite/unit-tests/clib/unicode_tests.ml` depend on?

#### The plan to wire the currently present unit tests to `dune` would then be thus:

* A `dune` file would be created in every sub-directory of `test-suite/unit-tests` that contains at least one `*.ml` file.
* Such a `dune` file would list all `*.ml` files in its directory within a `tests` stanza, allowing them to be run individually as well as all at once.
* It would also list the dependencies the change of which should trigger a re-run of the check.

I am going to try this out, see if it can fly and report.

**Comment by kindaro** [comment]
[See proof of concept in my fork.][1] You can try it out: change directory to `test-suite/unit-tests` and invoke `dune runtest`. There is one failing dummy check for the amusement of the curious reader.

On the down side, it is not feasible to support both `dune` and `make`, so this is a breaking change — I move files around and cut away unnecessary code. Also, the output of `dune runtest` is ugly — I opened an issue ocaml/dune#3937 to find out if it can be improved.

@ejgallego  Does this draft match your expectations?

[1]: https://github.com/coq/coq/compare/master...kindaro:feature-dune-unit-tests?expand=1

**Comment by kindaro** [comment]
> > * It is not trivial to provide _«quality of life»_ features such as colourful or machine-readable reporting.
> 
> I think the current quality of life status of that is not bad at all.

I do not know what your thinking is based on. [`dune` at present does not support any kind of report generation and if we want to have it, we shall have to patch it in ourselves:][1]

> This was discussed in the dune meeting and we acknowledge that this a problem. … We don't have the time to implement such a thing, but Arseniy & Cameron agreed to think about a possible design.

[1]: https://github.com/ocaml/dune/issues/3937#issuecomment-725570358

**Comment by ejgallego** [comment]
> See proof of concept in my fork.

Thanks! It is unfortunate we did this in parallel, but yesterday was a holiday in France so I didn't manage to submit.

I've started a PR to port not only the unit tests, but the full test-suite to dune, for now it doesn't include any modification but it still allows both schemes to work together. Please feel free to push / suggest modifications there.

https://github.com/coq/coq/pull/13364

> > I think the current quality of life status of that is not bad at all.
>
> I do not know what your thinking is based on. dune at present does not support any kind of report generation and if we want to have it, we shall have to patch it in ourselves:

Indeed it would be ideal to have nicer test setup in dune, and it will be eventually arrive. But what I mean is that for us, Coq devs, our current report generator works just fine, and indeed while we much welcome improvements, the pitfalls on the current testing setup make a very small part of other huge challenges we have in the development of Coq, such as performance, compatibility, and refactoring the compiler to make it incremental, not to speak of serious problems with some features.

Thus of all the problems we encounter in our daily dev life, having our small clunky and ugly reporting script is like 0.1% of them, hence the low priority. But I agree that indeed our optic is not very friendly for developers not touching hardcore parts of the system yet.

also cc: @rgrinberg

**Comment by kindaro** [comment]
Awesome! I was expecting it to take much longer. `dune runtest` from your branch runs very fast.

I have some notes though, if you are inclined to hear me out.

1. I see you have a much more elaborate system than what I drafted — I have no idea what all this OCaml code does. Is it really necessary? What are the design goals that it fulfills that simpler code would not?
2. Note that, since you are using custom `dune` rules instead of `test` or `tests`, `dune runtest --force` does nothing. Moreover, repeated runs in the presence of a failing test actually pass! It also means that whatever features associated with testing `dune` provides in the future would not be made use of.
3. Your approach is to discover tests to run at run time. This means that there is no _«master»_ list of tests to run — a test may go missing and no one will notice. My approach is to hard-wire every single test — that makes it less likely that some tests would disappear, as they have to also remove themselves from the `dune` file somehow.

Of course I allow it that I may be wrong throughout or simply irrelevant, so I absolutely understand if you ignore the above points. Thanks to your efforts, I can finally get back to doing what I was up to in the first place, so be assured of my gratitude.

**Comment by ejgallego** [comment]
> Awesome! I was expecting it to take much longer. `dune runtest` from your branch runs very fast.

There are only 4 unit tests, depending on the modules on the lower part so indeed it is fast.

> I have some notes though, if you are inclined to hear me out.

Of course.

> 1. I see you have a much more elaborate system than what I drafted — I have no idea what all this OCaml code does. Is it really necessary? What are the design goals that it fulfills that simpler code would not?

You mean the 45 lines OCaml program? It just generates the rules for the unit tests, it is more convenient to have a small generator like this. As you can see, we need to inform dune that the unit testing produces `.log` files.

> 2\. Note that, since you are using custom `dune` rules instead of `test` or `tests`, `dune runtest --force` does nothing. Moreover, repeated runs in the presence of a failing test actually pass! It also means that whatever features associated with testing `dune` provides in the future would not be made use of.

The `(tests ...)` stanza is just sugar for `(alias (name runtest) ...)` so indeed that is not the problem. Actually we may be running into https://github.com/ocaml/dune/issues/3415 ? That is a pretty serious bug in Dune if I understand correctly, but I did not have time to investigate.

> 3\. Your approach is to discover tests to run at run time. This means that there is no _«master»_ list of tests to run — a test may go missing and no one will notice. My approach is to hard-wire every single test — that makes it less likely that some tests would disappear, as they have to also remove themselves from the `dune` file somehow.

Note that the file with the tests is built and tracked in the version control system, and you have control over the generation.

Coq has many thousand test cases so I'm afraid keeping an explicit list is not an option, but a file like this can be reviewed. The problem you mention is also present in the make setup.

> 4\. Is it a requirement to keep supporting `make`? If the support may be dropped _(and it may be dropped)_, the code may be greatly simplified.

I would be interesting in hearing how, make support is going away soon we hope.

**Comment by ejgallego** [comment]
> The `(tests ...)` stanza is just sugar for `(alias (name runtest) ...)` so indeed that is not the problem. Actually we may be running into [ocaml/dune#3415](https://github.com/ocaml/dune/issues/3415) ? That is a pretty serious bug in Dune if I understand correctly, but I did not have time to investigate.

Umm, let me see if actually the problem is that we need to expect the right exit value.

**Comment by ejgallego** [comment]
> Umm, let me see if actually the problem is that we need to expect the right exit value.

Indeed, the failing test exists with `0`, so dune has no way to understand it has failed, that's a pitfall of our ounit setup, which is indeed badly in need of care.

**Comment by kindaro** [comment]
> > 4. Is it a requirement to keep supporting `make`? If the support may be dropped _(and it may be dropped)_, the code may be greatly simplified.
> 
> I would be interesting in hearing how, make support is going away soon we hope.

The answer is here:

> > 1. I see you have a much more elaborate system than what I drafted — I have no idea what all this OCaml code does. Is it really necessary? What are the design goals that it fulfills that simpler code would not?
> 
> … As you can see, we need to inform dune that the unit testing produces `.log` files.

— We can simplify by cutting away everything that pertains to `*.log` files. It should be the responsibility of the test runner to generate and format the reports. You can see an example in my draft — the diff has a lot of red, and much more can be cut where I merely replaced parameters with an underscore.

**Comment by kindaro** [comment]
> > Umm, let me see if actually the problem is that we need to expect the right exit value.
> 
> Indeed, the failing test exists with `0`, so dune has no way to understand it has failed, that's a pitfall of our ounit setup, which is indeed badly in need of care.

Actually this is fixed in my draft as well.

**Comment by ejgallego** [comment]
I see, would be indeed cool to integrate the parts of your proposal that can go now.

I still have the worry about the runner not being able to run test incrementally. That's a top priority.

**Comment by kindaro** [comment]
> > 3. Your approach is to discover tests to run at run time. This means that there is no _«master»_ list of tests to run — a test may go missing and no one will notice. My approach is to hard-wire every single test — that makes it less likely that some tests would disappear, as they have to also remove themselves from the `dune` file somehow.
> 
> Note that the file with the tests is built and tracked in the version control system, and you have control over the generation.
> 
> Coq has many thousand test cases so I'm afraid keeping an explicit list is not an option, but a file like this can be reviewed. The problem you mention is also present in the make setup.

I do not think one follows from another. We need to build a very long list of checks into a `dune` file once, but afterwards checks would be added one by one. Of course we should use a script to initialize the `dune` files — my proposition is that we should remove it afterwards, possibly by reverting a self-contained commit so that there is a trace in history. Keeping the script in the repository actually increases the cost of maintenance and risk:
* It increases complexity without any immediate gain. A `dune` file is easy to understand and has a narrow and predictable range of effects, while an OCaml script is hard to understand and may have any effect.
* When adding new checks, a script being around creates a temptation to re-generate the `dune` files with it instead of wiring the new checks in by hand. At this point a fault in the script may result in a section of checks disappearing from the list.

> I see, would be indeed cool to integrate the parts of your proposal that can go now.
> 
> I still have the worry about the runner not being able to run test incrementally. That's a top priority.

My understanding is that all 3 proposals we have for now are running incrementally? Or maybe I misunderstand the nature or criteria of your requirement?

**Comment by ejgallego** [comment]
> * When adding new checks, a script being around creates a temptation to re-generate the `dune` files with it instead of wiring the new checks in by hand. At this point a fault in the script may result in a section of checks disappearing from the list.

That's the whole point. If the script has bugs we will fix them, but it is still much easier to do a mistake on a manually maintained file. Not to say the current updating process is automatic and managed, and in your approach we would introduce another step in the test process for contributors.

> My understanding is that all 3 proposals we have for now are running incrementally? Or maybe I misunderstand the nature or criteria of your requirement?

Indeed, sorry. I think personally I'm pretty set on dune being the test runner, as it has a lot of nifty features, and above all does allow to run tests very quicky which is not the case today as it integrates automatically with the build system.

Also, I'd like to keep it simple and IMHO that does it. Furthermore, as improvements to Dune happen, we can benefit from them.

**Comment by kindaro** [comment]
> If the script has bugs we will fix them …

That is if you notice. I know you have code review, but you should not count on it because humans are not good at this sort of tasks. There was already an issue _(linked in the opening post of this issue)_ where unit tests were not discovered and no one noticed until I pointed out.

> … but it is still much easier to do a mistake on a manually maintained file.

Not if it is automatically checked.

* If you add a test and forget to wire it to dune, you will see that it is not failing — this is why the test driven development work flow requires one to first create a failing test. When following the proper work flow, it is impossible to forget to wire a check in.
* If later you accidentally remove the test file or otherwise make it inaccessible to `dune`, it will complain that a hard-wired task cannot be performed.

> Not to say the current updating process is automatic and managed, and in your approach we would introduce another step in the test process for contributors.

Which step? Regardless of whether new checks need to be wired in to `dune` files manually or those files get regenerated by a script, the updated `dune` files are committed, so there is a step to perform. It is simpler to add a few lines to a `dune` file by hand than to figure out how an arbitrarily complex script works and invoke it correctly.

I think maybe your mastery of OCaml and makes it difficult for you to see how much complexity you are adding by committing so much OCaml code. Your words:

> Also, I'd like to keep it simple and IMHO that does it.

— So do keep it simple!

That being said, this is not really an important question. I only aim to express my opinion precisely, not to make a case and contend for it.

* * *

> … I think personally I'm pretty set on dune being the test runner, as it has a lot of nifty features, and above all does allow to run tests very quicky which is not the case today as it integrates automatically with the build system.
>
> Also, I'd like to keep it simple and IMHO that does it. Furthermore, as improvements to Dune happen, we can benefit from them.

I absolutely agree.

**Comment by ejgallego** [comment]
> It is simpler to add a few lines to a `dune` file by hand than to figure out how an arbitrarily complex script works and invoke it correctly.

Actually the point is that you don't worry about that, the invocation and promotion are done automatically. The user just has to write the test, the rest works.
 
> I think maybe your mastery of OCaml and makes it difficult for you to see how much complexity you are adding by committing so much OCaml code.

I'm not sure 45 lines of code count as "too much code", but indeed IMHO it is not a big deal, would we decide to remove the script, switching to a manual updating setup would be easy.

**Comment by kindaro** [comment]
Is there any progress on this? Anything I can help with?

**Comment by ejgallego** [comment]
Hi @kindaro , thanks for you interest and help offer!

Indeed we have made some progress, last summary of the plan is in https://github.com/coq/coq/wiki/Coq-Call-2021-02-24 , notes.

I'd say that for now we are good, I think indeed the main problems today are checking that the documentations and plans are clear enough.

We also made quite a bit of progress on the new test-suite setup, while still not perfect you can check it out https://github.com/coq/coq/pull/13364 , should be functional for a set of tests.

**Comment by Zimmi48** [comment]
Thanks for writing these notes Emilio! Very good plan!
